// cashc/dist/ast/Location.js
export class Location {
    constructor(start, end) {
        this.start = start;
        this.end = end;
    }
    static fromCtx(ctx) {
        const stop = ctx.stop?.text ? ctx.stop : ctx.start;
        const textLength = (stop.text ?? '').length;
        const start = new Point(ctx.start.line, ctx.start.column);
        const end = new Point(stop.line, stop.column + textLength);
        return new Location(start, end);
    }
    static fromToken(token) {
        const textLength = (token.text ?? '').length;
        const start = new Point(token.line, token.column);
        const end = new Point(token.line, token.column + textLength);
        return new Location(start, end);
    }
    text(code) {
        return code.slice(this.start.offset(code), this.end.offset(code));
    }
}
export class Point {
    constructor(line, column) {
        this.line = line;
        this.column = column;
    }
    offset(code) {
        const lines = code.split('\n');
        const newLines = this.line - 1;
        const lineOffset = lines.slice(0, newLines).reduce((acc, curr) => acc + curr.length, 0);
        return lineOffset + newLines + this.column;
    }
    toString() {
        return `Line ${this.line}, Column ${this.column}`;
    }
}
