// cashc/dist/ast/Operator.js
export var NullaryOperator;
(function (NullaryOperator) {
    NullaryOperator["INPUT_INDEX"] = "this.activeInputIndex";
    NullaryOperator["BYTECODE"] = "this.activeBytecode";
    NullaryOperator["INPUT_COUNT"] = "tx.inputs.length";
    NullaryOperator["OUTPUT_COUNT"] = "tx.outputs.length";
    NullaryOperator["VERSION"] = "tx.version";
    NullaryOperator["LOCKTIME"] = "tx.locktime";
})(NullaryOperator || (NullaryOperator = {}));
export var UnaryOperator;
(function (UnaryOperator) {
    UnaryOperator["NOT"] = "!";
    UnaryOperator["NEGATE"] = "-";
    UnaryOperator["SIZE"] = ".length";
    UnaryOperator["REVERSE"] = ".reverse()";
    UnaryOperator["INPUT_VALUE"] = "tx.inputs[i].value";
    UnaryOperator["INPUT_LOCKING_BYTECODE"] = "tx.inputs[i].lockingBytecode";
    UnaryOperator["INPUT_OUTPOINT_HASH"] = "tx.inputs[i].outpointTransactionHash";
    UnaryOperator["INPUT_OUTPOINT_INDEX"] = "tx.inputs[i].outpointIndex";
    UnaryOperator["INPUT_UNLOCKING_BYTECODE"] = "tx.inputs[i].unlockingBytecode";
    UnaryOperator["INPUT_SEQUENCE_NUMBER"] = "tx.inputs[i].sequenceNumber";
    UnaryOperator["OUTPUT_VALUE"] = "tx.outputs[i].value";
    UnaryOperator["OUTPUT_LOCKING_BYTECODE"] = "tx.outputs[i].lockingBytecode";
    UnaryOperator["INPUT_TOKEN_CATEGORY"] = "tx.inputs[i].tokenCategory";
    UnaryOperator["INPUT_NFT_COMMITMENT"] = "tx.inputs[i].nftCommitment";
    UnaryOperator["INPUT_TOKEN_AMOUNT"] = "tx.inputs[i].tokenAmount";
    UnaryOperator["OUTPUT_TOKEN_CATEGORY"] = "tx.outputs[i].tokenCategory";
    UnaryOperator["OUTPUT_NFT_COMMITMENT"] = "tx.outputs[i].nftCommitment";
    UnaryOperator["OUTPUT_TOKEN_AMOUNT"] = "tx.outputs[i].tokenAmount";
})(UnaryOperator || (UnaryOperator = {}));
export var BinaryOperator;
(function (BinaryOperator) {
    BinaryOperator["MUL"] = "*";
    BinaryOperator["DIV"] = "/";
    BinaryOperator["MOD"] = "%";
    BinaryOperator["PLUS"] = "+";
    BinaryOperator["MINUS"] = "-";
    BinaryOperator["LT"] = "<";
    BinaryOperator["LE"] = "<=";
    BinaryOperator["GT"] = ">";
    BinaryOperator["GE"] = ">=";
    BinaryOperator["EQ"] = "==";
    BinaryOperator["NE"] = "!=";
    BinaryOperator["BIT_AND"] = "&";
    BinaryOperator["BIT_XOR"] = "^";
    BinaryOperator["BIT_OR"] = "|";
    BinaryOperator["AND"] = "&&";
    BinaryOperator["OR"] = "||";
    BinaryOperator["SPLIT"] = ".split";
})(BinaryOperator || (BinaryOperator = {}));
