// cashc/dist/generation/GenerateTargetTraversal.js
import { hexToBin } from '@bitauth/libauth';
import { asmToScript, encodeBool, encodeInt, encodeString, Op, PrimitiveType, resultingType, scriptToAsm, } from '@cashscript/utils';
import { HexLiteralNode, } from '../ast/AST.js';
import AstTraversal from '../ast/AstTraversal.js';
import { GlobalFunction, Class } from '../ast/Globals.js';
import { BinaryOperator } from '../ast/Operator.js';
import { compileBinaryOp, compileCast, compileGlobalFunction, compileNullaryOp, compileTimeOp, compileUnaryOp, } from './utils.js';
export default class GenerateTargetTraversal extends AstTraversal {
    constructor() {
        super(...arguments);
        this.output = [];
        this.stack = [];
        this.scopeDepth = 0;
    }
    emit(op) {
        if (Array.isArray(op)) {
            this.output.push(...op);
        }
        else {
            this.output.push(op);
        }
    }
    pushToStack(value, pushToBottom) {
        if (pushToBottom) {
            this.stack.push(value);
        }
        else {
            this.stack.unshift(value);
        }
    }
    popFromStack(count = 1) {
        for (let i = 0; i < count; i += 1) {
            this.stack.shift();
        }
    }
    removeFromStack(i) {
        this.stack.splice(i, 1);
    }
    nipFromStack() {
        this.stack.splice(1, 1);
    }
    getStackIndex(value) {
        const index = this.stack.indexOf(value);
        if (index === -1)
            throw new Error(); // Should not happen
        return index;
    }
    visitSourceFile(node) {
        node.contract = this.visit(node.contract);
        // Minimally encode output by going Script -> ASM -> Script
        this.output = asmToScript(scriptToAsm(this.output));
        return node;
    }
    visitContract(node) {
        node.parameters = this.visitList(node.parameters);
        if (node.functions.length === 1) {
            node.functions = this.visitList(node.functions);
        }
        else {
            this.pushToStack('$$', true);
            node.functions = node.functions.map((f, i) => {
                const stackCopy = [...this.stack];
                const selectorIndex = this.getStackIndex('$$');
                this.emit(encodeInt(BigInt(selectorIndex)));
                if (i === node.functions.length - 1) {
                    this.emit(Op.OP_ROLL);
                    this.removeFromStack(selectorIndex);
                }
                else {
                    this.emit(Op.OP_PICK);
                }
                // All functions are if-else statements, except the final one which is
                // enforced with NUMEQUALVERIFY
                this.emit(encodeInt(BigInt(i)));
                this.emit(Op.OP_NUMEQUAL);
                if (i < node.functions.length - 1) {
                    this.emit(Op.OP_IF);
                }
                else {
                    this.emit(Op.OP_VERIFY);
                }
                f = this.visit(f);
                if (i < node.functions.length - 1) {
                    this.emit(Op.OP_ELSE);
                }
                this.stack = [...stackCopy];
                return f;
            });
            for (let i = 0; i < node.functions.length - 1; i += 1) {
                this.emit(Op.OP_ENDIF);
            }
        }
        return node;
    }
    visitFunctionDefinition(node) {
        this.currentFunction = node;
        node.parameters = this.visitList(node.parameters);
        node.body = this.visit(node.body);
        this.removeFinalVerify();
        this.cleanStack();
        return node;
    }
    removeFinalVerify() {
        // After EnsureFinalRequireTraversal, we know that the final opcodes are either
        // "OP_VERIFY", "OP_CHECK{LOCKTIME|SEQUENCE}VERIFY OP_DROP" or "OP_ENDIF"
        const finalOp = this.output.pop();
        // If the final op is OP_VERIFY and the stack size is less than 4 we remove it from the script
        // - We have the stack size check because it is more efficient to use 2DROP rather than NIP
        //   if >= 4 elements are left (5 including final value) (e.g. 2DROP 2DROP 1 < NIP NIP NIP NIP)
        if (finalOp === Op.OP_VERIFY && this.stack.length < 4) {
            // Since the final value is no longer popped from the stack by OP_VERIFY,
            // we add it back to the stack
            this.pushToStack('(value)');
        }
        else {
            this.emit(finalOp);
            // At this point there is no verification value left on the stack:
            //  - scoped stack is cleared inside branch ended by OP_ENDIF
            //  - OP_CHECK{LOCKTIME|SEQUENCE}VERIFY OP_DROP does not leave a verification value
            // so we add OP_1 to the script (indicating success)
            this.emit(Op.OP_1);
            this.pushToStack('(value)');
        }
    }
    cleanStack() {
        // Keep final verification value, OP_NIP the other stack values
        const stackSize = this.stack.length;
        for (let i = 0; i < stackSize - 1; i += 1) {
            this.emit(Op.OP_NIP);
            this.nipFromStack();
        }
    }
    visitParameter(node) {
        this.pushToStack(node.name, true);
        return node;
    }
    visitVariableDefinition(node) {
        node.expression = this.visit(node.expression);
        this.popFromStack();
        this.pushToStack(node.name);
        return node;
    }
    visitTupleAssignment(node) {
        node.tuple = this.visit(node.tuple);
        this.popFromStack(2);
        this.pushToStack(node.var1.name);
        this.pushToStack(node.var2.name);
        return node;
    }
    visitAssign(node) {
        node.expression = this.visit(node.expression);
        if (this.scopeDepth > 0) {
            this.emitReplace(this.getStackIndex(node.identifier.name));
            this.popFromStack();
        }
        else {
            this.popFromStack();
            this.pushToStack(node.identifier.name);
        }
        return node;
    }
    // This algorithm can be optimised for hardcoded depths
    // See thesis for explanation
    emitReplace(index) {
        this.emit(encodeInt(BigInt(index)));
        this.emit(Op.OP_ROLL);
        this.emit(Op.OP_DROP);
        for (let i = 0; i < index - 1; i += 1) {
            this.emit(Op.OP_SWAP);
            if (i < index - 2) {
                this.emit(Op.OP_TOALTSTACK);
            }
        }
        for (let i = 0; i < index - 2; i += 1) {
            this.emit(Op.OP_FROMALTSTACK);
        }
    }
    visitTimeOp(node) {
        node.expression = this.visit(node.expression);
        this.emit(compileTimeOp(node.timeOp));
        this.popFromStack();
        return node;
    }
    visitRequire(node) {
        node.expression = this.visit(node.expression);
        this.emit(Op.OP_VERIFY);
        this.popFromStack();
        return node;
    }
    visitBranch(node) {
        node.condition = this.visit(node.condition);
        this.popFromStack();
        this.scopeDepth += 1;
        this.emit(Op.OP_IF);
        let stackDepth = this.stack.length;
        node.ifBlock = this.visit(node.ifBlock);
        this.removeScopedVariables(stackDepth);
        if (node.elseBlock) {
            this.emit(Op.OP_ELSE);
            stackDepth = this.stack.length;
            node.elseBlock = this.visit(node.elseBlock);
            this.removeScopedVariables(stackDepth);
        }
        this.emit(Op.OP_ENDIF);
        this.scopeDepth -= 1;
        return node;
    }
    removeScopedVariables(depthBeforeScope) {
        const dropCount = this.stack.length - depthBeforeScope;
        for (let i = 0; i < dropCount; i += 1) {
            this.emit(Op.OP_DROP);
            this.popFromStack();
        }
    }
    visitCast(node) {
        node.expression = this.visit(node.expression);
        // Special case for sized bytes cast, since it has another node to traverse
        if (node.size) {
            node.size = this.visit(node.size);
            this.emit(Op.OP_NUM2BIN);
            this.popFromStack();
        }
        this.emit(compileCast(node.expression.type, node.type));
        this.popFromStack();
        this.pushToStack('(value)');
        return node;
    }
    visitFunctionCall(node) {
        if (node.identifier.name === GlobalFunction.CHECKMULTISIG) {
            return this.visitMultiSig(node);
        }
        node.parameters = this.visitList(node.parameters);
        this.emit(compileGlobalFunction(node.identifier.name));
        this.popFromStack(node.parameters.length);
        this.pushToStack('(value)');
        return node;
    }
    visitMultiSig(node) {
        this.emit(encodeBool(false));
        this.pushToStack('(value)');
        node.parameters = this.visitList(node.parameters);
        this.emit(Op.OP_CHECKMULTISIG);
        const sigs = node.parameters[0];
        const pks = node.parameters[1];
        this.popFromStack(sigs.elements.length + pks.elements.length + 3);
        this.pushToStack('(value)');
        return node;
    }
    visitInstantiation(node) {
        if (node.identifier.name === Class.LOCKING_BYTECODE_P2PKH) {
            // OP_DUP OP_HASH160 OP_PUSH<20>
            this.emit(hexToBin('76a914'));
            this.pushToStack('(value)');
            // <pkh>
            this.visit(node.parameters[0]);
            this.emit(Op.OP_CAT);
            // OP_EQUAL OP_CHECKSIG
            this.emit(hexToBin('88ac'));
            this.emit(Op.OP_CAT);
            this.popFromStack(2);
        }
        else if (node.identifier.name === Class.LOCKING_BYTECODE_P2SH20) {
            // OP_HASH160 OP_PUSH<20>
            this.emit(hexToBin('a914'));
            this.pushToStack('(value)');
            // <script hash>
            this.visit(node.parameters[0]);
            this.emit(Op.OP_CAT);
            // OP_EQUAL
            this.emit(hexToBin('87'));
            this.emit(Op.OP_CAT);
            this.popFromStack(2);
        }
        else if (node.identifier.name === Class.LOCKING_BYTECODE_P2SH32) {
            // OP_HASH256 OP_PUSH<32>
            this.emit(hexToBin('aa20'));
            this.pushToStack('(value)');
            // <script hash>
            this.visit(node.parameters[0]);
            this.emit(Op.OP_CAT);
            // OP_EQUAL
            this.emit(hexToBin('87'));
            this.emit(Op.OP_CAT);
            this.popFromStack(2);
        }
        else if (node.identifier.name === Class.LOCKING_BYTECODE_NULLDATA) {
            // Total script = OP_RETURN (<VarInt> <chunk>)+
            // OP_RETURN
            this.emit(hexToBin('6a'));
            this.pushToStack('(value)');
            const { elements } = node.parameters[0];
            // <VarInt data chunk size (dynamic)>
            elements.forEach((el) => {
                this.visit(el);
                // Push the element's size (and calculate VarInt)
                this.emit(Op.OP_SIZE);
                if (el instanceof HexLiteralNode) {
                    // If the argument is a literal, we know its size
                    if (el.value.byteLength > 75) {
                        this.emit(hexToBin('4c'));
                        this.emit(Op.OP_SWAP);
                        this.emit(Op.OP_CAT);
                    }
                }
                else {
                    // If the argument is not a literal, the script needs to check size
                    this.emit(Op.OP_DUP);
                    this.emit(encodeInt(75n));
                    this.emit(Op.OP_GREATERTHAN);
                    this.emit(Op.OP_IF);
                    this.emit(hexToBin('4c'));
                    this.emit(Op.OP_SWAP);
                    this.emit(Op.OP_CAT);
                    this.emit(Op.OP_ENDIF);
                }
                // Concat size and arguments
                this.emit(Op.OP_SWAP);
                this.emit(Op.OP_CAT);
                this.emit(Op.OP_CAT);
                this.popFromStack();
            });
            this.popFromStack();
        }
        else {
            throw new Error(); // Should not happen
        }
        this.pushToStack('(value)');
        return node;
    }
    visitTupleIndexOp(node) {
        node.tuple = this.visit(node.tuple);
        if (node.index === 0) {
            this.emit(Op.OP_DROP);
            this.popFromStack();
        }
        else if (node.index === 1) {
            this.emit(Op.OP_NIP);
            this.nipFromStack();
        }
        return node;
    }
    visitBinaryOp(node) {
        node.left = this.visit(node.left);
        node.right = this.visit(node.right);
        const isNumeric = resultingType(node.left.type, node.right.type) === PrimitiveType.INT;
        this.emit(compileBinaryOp(node.operator, isNumeric));
        this.popFromStack(2);
        this.pushToStack('(value)');
        if (node.operator === BinaryOperator.SPLIT)
            this.pushToStack('(value)');
        return node;
    }
    visitUnaryOp(node) {
        node.expression = this.visit(node.expression);
        this.emit(compileUnaryOp(node.operator));
        this.popFromStack();
        this.pushToStack('(value)');
        return node;
    }
    visitNullaryOp(node) {
        this.emit(compileNullaryOp(node.operator));
        this.pushToStack('(value)');
        return node;
    }
    visitArray(node) {
        node.elements = this.visitList(node.elements);
        this.emit(encodeInt(BigInt(node.elements.length)));
        this.pushToStack('(value)');
        return node;
    }
    visitIdentifier(node) {
        const stackIndex = this.getStackIndex(node.name);
        this.emit(encodeInt(BigInt(stackIndex)));
        // If the final use is inside an if-statement, we still OP_PICK it
        // We do this so that there's no difference in stack depths between execution paths
        if (this.isOpRoll(node)) {
            this.emit(Op.OP_ROLL);
            this.removeFromStack(stackIndex);
        }
        else {
            this.emit(Op.OP_PICK);
        }
        this.pushToStack('(value)');
        return node;
    }
    isOpRoll(node) {
        return this.currentFunction.opRolls.get(node.name) === node && this.scopeDepth === 0;
    }
    visitBoolLiteral(node) {
        this.emit(encodeBool(node.value));
        this.pushToStack('(value)');
        return node;
    }
    visitIntLiteral(node) {
        this.emit(encodeInt(node.value));
        this.pushToStack('(value)');
        return node;
    }
    visitStringLiteral(node) {
        this.emit(encodeString(node.value));
        this.pushToStack('(value)');
        return node;
    }
    visitHexLiteral(node) {
        this.emit(node.value);
        this.pushToStack('(value)');
        return node;
    }
}
